05_testing_and_ci/python_testing_demo.md

```diff
@@ -0,0 +1,58 @@
+- The regression test first reads an old data set and an average value from a CSV file. Then the function `find_average` is run with the old data set and the new average value is compared to the old one.
+
+## pytest
+
+- pytest is installed by: `pip install pytest`.
+- All tests can be run using the command-line tool called `pytest`.
+- **Comparing floating point variables** needs to be handled in functions like `find_average` and is done using `pytest.approx(value, abs)`.
+- Even if one test fails, pytest runs all the tests and gives a report on the failing test. The assertion failure report generated my pytest is also more detailed than the usual Python assertion report.
+- pytest can run unittest suites directly using `pytest tests`.
+- pytest is able to detect tests in several forms of folder structures, and the folder structures have advantages and disadvantages. More information on this is in the [documentation](https://docs.pytest.org/en/6.2.x/goodpractices.html#choosing-a-test-layout-import-rules).
+- Example:
+
+```
+setup.py
+my_software/
+    __init__.py
+    main.py
+    feature.py
+    test/
+        __init__.py
+        test_one.py
+        test_two.py
+        ...
+```
+
+## unittest
+
+- Base class `unittest.TestCase` is used to create a test suite consisting of all the tests of a software.
+- Each test is now a function of a class which is derived from the class `unittest.TestCase`.
+- The same tests are implemented using `unittest` in the file `test_operations_unittests.py` as functions of a class `TestOperations`.
+- unittest.TestCase offers functions like `assertEqual`, `assertAlmostEqual`, `assertTrue`, etc. for use instead of the usual assertion statements. These statements help the test runner to accumulate all test results and generate a test report.
+- `unittest.main()` provides an option to run the tests from a command-line interface.
+- `setUp` function in run before all the tests. Similar a clean up function `tearDown` exists.
+- The intention is to group together sets of similar tests in an instant of `unittest.TestCase` and have multiple such instances.
+- Decorators such as `@unittest.skip`, `@unittest.skipIf`, `@unittest.expectedFailure` can be used to gain flexibility over working of tests.
+- `unittest.TestCase.subTest` can be used to distinguish parameters inside the body of a test.
+
+## coverage
+
+- Installing coverage using pip: `pip install coverage`.
+- Testing frameworks can be run via coverage. Lets take our first example and run pytest via coverage: `coverage run -m pytest`.
+- coverage does not generate any output immediately as it would interfere with the test output.
+- Code coverage information is stored in a file `.coverage` in the working directory. This information can be viewed using `coverage report -m`
+- A more fancier report can be viewed by generating html output using `coverage html`.
+- The file `htmlcov/index.html` can be opened in a browser to view the test coverage report.
```

Comment:
- The coverage section is very nice to reproduce. Here it is clear which commands to use and what they do. Maybe one should mention (again) in which directory one should be when one runs the command. You could still give some sample output (or parts of it) such that it is easier to understand whether one got the expected result.
- Generic warning: "No new file at end of file"